python_ta/checkers/unnecessary_assignment_checker.py

```diff
@@ -0,0 +1,99 @@
+"""
+Checker for unnecessary assignment to a variable.
+"""
+
+import astroid
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import check_messages
+
+
+class UnnecessaryAssignmentChecker(BaseChecker):
+    """Checker for unnecessary assignment to variables in several cases."""
+    __implements__ = IAstroidChecker
+    # name is the same as file name but without _checker part
+    name = 'unnecessary_assignment'
+    # use dashes for connecting words in message symbol. C for the message-id is because
+    #  I believe this to be a convention issue.
+    msgs = {'E9917': ('There is an unnecessary assignment within the function',
+                      'unnecessary-assignment',
+                      'Used when there is an assignment that does not affect the '
+                      'function\'s return value'),
+            }
+
+    # this is important so that your checker is executed before others. It seems all
+    # the checkers have the same priority though.
+    priority = -1
+
+    def _check_unnecessary_assignment(self, node):
+        """Return whether this function has an instance of unnecessary assignment."""
+
+        values = {}  # creates a 'notetaker' for when values are assigned and used.
+        # the values of the keys in values will be a list where the first element is a number
+        # 0 : assigned too but unused
+        # -1 : used
+        # >1 : assigned to before use.
+        # and the second element the last assignment node,
+        # which can be referenced as the unnecessary one if assigned to again before use.
+        # the third element is whether it has been seen in the function before.
```

Comment:
I suspect that this third element is not necessary (you mentioned it in the presentation today, but I didn't quite get what you meant). But anyway, here is a suggestion for how to simplify this data structure:

For each name, store *a single node, which is either an `AssignName` or a `Name` node*. Then rather than checking if `values[block.name][0] > 1`, you can check `isinstance(values[block.name], astroid.AssignName)` (and so on for other changes).